docs/spec/concepts.rst

```diff
@@ -3,97 +3,186 @@
 Type system concepts
 ====================
 
-.. _`union-types`:
+Static, dynamic, and gradual typing
+-----------------------------------
+
+A **statically-typed** programming language runs a type-checker before running
+a program. The program is required to be well-typed according to the language's
+type system. The type system assigns a type to all expressions in the language
+and verifies that their uses obey the typing rules. Normally, a program that is
+not well-typed (i.e., one that contains a type error) will not run. Java and
+C++ are examples of statically-typed object-oriented languages.
+
+A **dynamically-typed** programming language does not run a type-checker before
+running a program. Instead, it will check the types of values before performing
+operations on them at runtime. This is not to say that the language is
+"untyped". Values at runtime have a type and their uses obey typing rules. Not
+every operation will be checked, but certain primitive operations in the
+language such as attribute access or arithmetic will be. Python was
+historically a dynamically-typed language.
+
+**Gradual typing** is the name for a specific way to combine static and dynamic
+typing. It allows mixing static and dynamic checking at a fine level of
+granularity. For instance, in type-annotated Python, individual variables,
+parameters, and returns can optionally be given a static type. Data structures
+such as objects can have a mix of static and dynamic-only checking. As an
+example, a Python dictionary could be annotated to have static checking of the
+key type but only dynamic checking of the value type.
+
+In gradual typing, a dynamically typed value is indicated by a special
```

Comment:
Since here, we are are describing the behavior of the `dyn` type with respect to how it gets type checked, could we say something about the design philosophy behind the typechecker behavior? 

For example, the reason expressions that are annotated with `dyn` are not be subject to static type-checking, is that we optimistically assume that that the value for those expressions will have appropriate types at runtime.

Reply:
Right, we could also discuss the gradual guarantee as a way to motivate this behavior. I guess essentially both framings of the idea boil down to the assumption that we are adding static types to an already-working untyped program.

Reply:
I'd be careful not to frame this misleadingly.  It kind of sounds like the programmer should use `Any` for code that they believe is correct.

A lot of gradual typing systems will insert runtime checked casts of dynamic values, explicitly not assuming that they are correct.

There are all kinds of reasons that Python programmers might write new code with `Any`.

Reply:
I added a short paragraph on the gradual guarantee. I'm not sure we need it, but it's probably useful to mention.